File: 백트래킹/leetcode131(Medium)-Palindrome Partitioning.js
// 백트래킹 : 284ms(60.85%), 63.5MB(64.95%)
/**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function (s) {
  const N = s.length
  const result = []

  const isPalindrome = (string) => {
    return string === [...string].reverse().join("")
  }

  const dfs = (start, array) => {
    if (start >= N) {
      result.push(array)
      return
    }

    for (let end = start; end < N; end++) {
      const substr = s.slice(start, end + 1)
      if (isPalindrome(substr)) {
        array.push(substr)
        dfs(end + 1, [...array])
        array.pop()
      }
    }
  }

  dfs(0, [])

  return result
}

// 백트래킹 + DP : 248ms(85.89%), 64.4MB(57.97%)
/**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function (s) {
  const N = s.length
  const result = []
  const dp = Array.from({ length: N }, () => new Array(N).fill(false))

  const dfs = (start, array) => {
    if (start >= N) {
      result.push(array)
      return
    }

    for (let end = start; end < N; end++) {
      if (s[start] === s[end] && (end - start <= 2 || dp[start + 1][end - 1])) {
        dp[start][end] = true
        array.push(s.slice(start, end + 1))
        dfs(end + 1, [...array])
        array.pop()
      }
    }
  }

  dfs(0, [])

  return result
}

File: 백트래킹/leetcode131(Medium)-Palindrome Partitioning.py
# 백트래킹으로만 풀 경우 : O(n * 2^n) 706ms(53.40)%, 30.3MB(62.53%)
class Solution:
    def is_palindrome(self, s):
        return s == s[::-1]

    def partition(self, s: str) -> List[List[str]]:
        result = []

        def dfs(start, array):
            if start >= len(s):
                result.append(array)
                return
                
            for end in range(start, len(s)):
                substr = s[start: end + 1]
                if self.is_palindrome(substr):
                    array.append(substr)
                    dfs(end + 1, array[:])
                    array.pop()

        dfs(0, [])

        return result

# 백트래킹 + DP로 최적화: O(n^2), 652ms(88.97%), 30.1MB(69.72%)
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        N = len(s)
        
        result = []

        dp = [[False] * N for _ in range(N)]

        def dfs(start, array):
            nonlocal dp
            if start >= N:
                result.append(array)
                return

            for end in range(start, N):
                if s[start] == s[end] and (end - start <= 2 or dp[start + 1][end - 1]):
                    dp[start][end] = True
                    array.append(s[start : end + 1])
                    dfs(end + 1, array[:])
                    array.pop()

        dfs(0, [])

        return result